File: contracts/social-token/reference/social-token-go/AMM.go
package socialtoken

import (
	"errors"
	"math"

	"github.com/shopspring/decimal"
)

/* -------------------------------------------------------------------------------------------------
integral: this function determines the integral of the AMM curve.
------------------------------------------------------------------------------------------------- */

func integral(AMM string, upperBound decimal.Decimal, lowerBound decimal.Decimal,
	targetPrice decimal.Decimal, targetSupply decimal.Decimal) (decimal.Decimal, error) {
	switch AMM {

	case LINEAR_AMM:
		multiplier := targetPrice.Div(targetSupply)
		integral, err := saveSubstraction(upperBound.Pow(TWO_DECIMAL), lowerBound.Pow(TWO_DECIMAL))
		return multiplier.Mul(integral).Div(TWO_DECIMAL), err

	case QUADRATIC_AMM:
		multiplier := targetPrice.Div(targetSupply.Pow(TWO_DECIMAL))
		integral, err := saveSubstraction(upperBound.Pow(THREE_DECIMAL), lowerBound.Pow(THREE_DECIMAL))
		return multiplier.Mul(integral).Div(THREE_DECIMAL), err

	case EXPONENTIAL_AMM:
		multiplier := targetPrice.Mul(Exponent(EXP_DECIMAL, targetSupply.Neg()))
		integral, err := saveSubstraction(Exponent(EXP_DECIMAL, upperBound), Exponent(EXP_DECIMAL, lowerBound))
		return multiplier.Mul(integral), err

	case SIGMOID_AMM:
		// upper := (upperBound + math.Log(1+math.Exp(-upperBound+targetSupply)))
		upperExpFloat, _ := (Exponent(EXP_DECIMAL, (upperBound.Neg().Add(targetSupply))).Add(ONE_DECIMAL)).Float64()
		upper := (decimal.NewFromFloat(math.Log(upperExpFloat))).Add(upperBound)
		// lower := (lowerBound + math.Log(1+math.Exp(-lowerBound+targetSupply)))
		lowerExpFloat, _ := (Exponent(EXP_DECIMAL, (lowerBound.Neg().Add(targetSupply))).Add(ONE_DECIMAL)).Float64()
		lower := (decimal.NewFromFloat(math.Log(lowerExpFloat))).Add(lowerBound)
		integral, err := saveSubstraction(upper, lower)
		return targetPrice.Div(TWO_DECIMAL).Mul(integral), err
	}
	return decimal.Zero, errors.New("ERROR COMPUTING THE INTEGRAL. ")
}

/* -------------------------------------------------------------------------------------------------
marketPrice: this function determines the market price.
------------------------------------------------------------------------------------------------- */

func getMarketPrice(AMM string, supplyReleased decimal.Decimal, initialSupply decimal.Decimal,
	targetPrice decimal.Decimal, targetSupply decimal.Decimal) (decimal.Decimal, error) {

	effectiveSupply := decimal.Max(decimal.Zero, supplyReleased.Sub(initialSupply))
	// if err != nil {
	// 	return decimal.Zero, err
	// }

	switch AMM {
	case LINEAR_AMM:
		multiplier := targetPrice.Div(targetSupply)
		return multiplier.Mul(effectiveSupply), nil

	case QUADRATIC_AMM:
		multiplier := targetPrice.Div(targetSupply.Pow(TWO_DECIMAL))
		return multiplier.Mul(effectiveSupply.Pow(TWO_DECIMAL)), nil

	case EXPONENTIAL_AMM:
		multiplier := targetPrice.Mul(Exponent(EXP_DECIMAL, targetSupply.Neg()))
		return multiplier.Mul(EXP_DECIMAL.Pow(supplyReleased)), nil

	case SIGMOID_AMM:
		// 	return targetPrice * (1. / (1 + math.Exp(-effectiveSupply+targetSupply))), nil
		return targetPrice.Mul(ONE_DECIMAL.Div(Exponent(EXP_DECIMAL, (effectiveSupply.Neg().Add(targetSupply))).Add(ONE_DECIMAL))), nil
	}
	return decimal.Zero, errors.New("ERROR COMPUTING GETTING THE MARKET PRICE. ")
}

/* -------------------------------------------------------------------------------------------------
buyingSocialTokens: this function determines the amount of X of Funding Tokens to receive after an
                 investment of Y Pod Tokens
------------------------------------------------------------------------------------------------- */
func buyingSocialTokens(AMM string, supplyReleased decimal.Decimal, initialSupply decimal.Decimal,
	amount decimal.Decimal, targetPrice decimal.Decimal, targetSupply decimal.Decimal) (decimal.Decimal, error) {
	effectiveSupply := decimal.Max(decimal.Zero, supplyReleased.Sub(initialSupply))
	//effectiveSupply, err := saveSubstraction(supplyReleased, initialSupply)
	// if err != nil {
	// 	return decimal.Zero, err
	// }
	newSupply := effectiveSupply.Add(amount)
	fundingAmount, err := integral(AMM, newSupply, effectiveSupply, targetPrice, targetSupply)
	return fundingAmount, err
}

/* -------------------------------------------------------------------------------------------------
selling_social_tokens: this function determines the amount of X of Funding Tokens to give after
               selling Y Funding Tokens
------------------------------------------------------------------------------------------------- */
func selling_social_tokens(AMM string, supplyReleased decimal.Decimal, initialSupply decimal.Decimal,
	sellingAmount decimal.Decimal, spread decimal.Decimal, targetPrice decimal.Decimal, targetSupply decimal.Decimal) (decimal.Decimal, error) {

	// Compute supply left after selling it //
	effectiveSupply := decimal.Max(decimal.Zero, supplyReleased.Sub(initialSupply))
	// effectiveSupply, err := saveSubstraction(supplyReleased, initialSupply)
	// if err != nil {
	// 	return decimal.Zero, err
	// }
	lowSupply := decimal.Max(decimal.Zero, effectiveSupply.Div(sellingAmount))

	fundingAmount, err := integral(AMM, effectiveSupply, lowSupply, targetPrice, targetSupply)
	if err != nil {
		return decimal.Zero, err
	}
	return fundingAmount.Mul(ONE_DECIMAL.Sub(spread)), err
}

/* -------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------- */

File: contracts/social-token/src/amm.rs
//! Functions related to automated market maker

use super::*;
#[cfg(feature = "std")]
use ink_storage::traits::StorageLayout;
use ink_storage::traits::{PackedLayout, SpreadLayout};
use rust_decimal::prelude::*;

/// Type of bonding curve
#[derive(Debug, PartialEq, Eq, Encode, Decode, SpreadLayout, PackedLayout, Copy, Clone)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo, StorageLayout))]
pub enum AmmType {
    Linear,
    Quadratic,
    Exponential,
    Sigmoid,
}

/// Calculate the integral of the AMM curve.
pub fn integral(
    amm_type: AmmType,
    upper_bound: Decimal,
    lower_bound: Decimal,
    target_price: Decimal,
    target_supply: Decimal,
) -> Result<Decimal> {
    /// Subtract and return an error if first argument becomes negative
    fn substract(main: Decimal, amount: Decimal) -> Result<Decimal> {
        let main = main - amount;
        if main.is_sign_negative() {
            return Err(Error::InsufficientBalance);
        }
        Ok(main)
    }

    match amm_type {
        AmmType::Linear => {
            let multiplier = target_price / target_supply;
            let integral = substract(upper_bound.powu(2), lower_bound.powu(2))?;
            Ok(multiplier * integral * dec!(0.5))
        }
        AmmType::Quadratic => {
            let multiplier = (target_price / target_supply).powu(2);
            let integral = substract(upper_bound.powu(3), lower_bound.powu(3))?;
            Ok((multiplier * integral) / dec!(3))
        }
        AmmType::Exponential => {
            let multiplier = target_price * (-target_supply).exp();
            let integral = substract(upper_bound.exp(), lower_bound.exp())?;
            Ok(multiplier * integral)
        }
        AmmType::Sigmoid => {
            let upper = ((target_supply - upper_bound).exp() + dec!(1)).ln() + upper_bound;
            let lower = ((target_supply - lower_bound).exp() + dec!(1)).ln() + lower_bound;
            let integral = substract(upper, lower)?;
            Ok(target_price * dec!(0.5) * integral)
        }
    }
}

// /// Calculate the market price
// pub fn get_market_price(
//     amm_type: AmmType,
//     supply_released: Decimal,
//     initial_supply: Decimal,
//     target_price: Decimal,
//     target_supply: Decimal,
// ) -> Result<Decimal> {
//     let effective_supply = dec!(0).max(supply_released - initial_supply);
//
//     match amm_type {
//         AmmType::Linear => Ok((target_price / target_supply) * effective_supply),
//         AmmType::Quadratic => Ok((target_price / target_supply.powu(2)) * effective_supply.powu(2)),
//         AmmType::Exponential => Ok((target_price * (-target_supply).exp()) * supply_released.exp()),
//         AmmType::Sigmoid => Ok(target_price * (dec!(1) / ((target_supply - effective_supply).exp() + dec!(1)))),
//     }
// }

/// Determines the amount of X of Funding Tokens to receive after an investment of Y Pod Tokens
pub fn price_for_mint(
    amm_type: AmmType,
    supply_released: Decimal,
    initial_supply: Decimal,
    amount: Decimal,
    target_price: Decimal,
    target_supply: Decimal,
) -> Result<Decimal> {
    let effective_supply = dec!(0).max(supply_released - initial_supply);

    let new_supply = effective_supply + amount;
    integral(amm_type, new_supply, effective_supply, target_price, target_supply)
}

/// Determines the amount of X of Funding Tokens to give after selling Y Funding Tokens
pub fn reward_for_burn(
    amm_type: AmmType,
    supply_released: Decimal,
    initial_supply: Decimal,
    selling_amount: Decimal,
    target_price: Decimal,
    target_supply: Decimal,
) -> Result<Decimal> {
    // Compute supply left after selling it //
    let effective_supply = dec!(0).max(supply_released - initial_supply);
    let low_supply = dec!(0).max(effective_supply / selling_amount);

    integral(amm_type, effective_supply, low_supply, target_price, target_supply)
}
